packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js

```diff
@@ -199,124 +202,143 @@ export function BlockSettingsDropdown( {
+									/>
+									{ shouldShowBlockParentMenuItem && (
 										<BlockParentSelectorMenuItem
 											parentClientId={
 												firstParentClientId
 											}
 											parentBlockType={ parentBlockType }
 										/>
 									) }
-								{ count === 1 && (
-									<BlockHTMLConvertButton
-										clientId={ firstBlockClientId }
-									/>
-								) }
-								{ ! isContentOnly && (
-									<CopyMenuItem
-										clientIds={ clientIds }
-										onCopy={ onCopy }
-										shortcut={ displayShortcut.primary(
-											'c'
-										) }
-									/>
-								) }
-								{ canDuplicate && (
-									<MenuItem
-										onClick={ pipe(
-											onClose,
-											onDuplicate,
-											updateSelectionAfterDuplicate
-										) }
-										shortcut={ shortcuts.duplicate }
-									>
-										{ __( 'Duplicate' ) }
-									</MenuItem>
-								) }
-								{ canInsertBlock && ! isContentOnly && (
-									<>
+									{ count === 1 && (
+										<BlockHTMLConvertButton
+											clientId={ firstBlockClientId }
+										/>
+									) }
+									{ ! isContentOnly && (
+										<CopyMenuItem
+											clientIds={ clientIds }
+											onCopy={ onCopy }
+											shortcut={ displayShortcut.primary(
+												'c'
+											) }
+										/>
+									) }
+									{ canDuplicate && (
 										<MenuItem
 											onClick={ pipe(
 												onClose,
-												onInsertBefore
+												onDuplicate,
+												updateSelectionAfterDuplicate
 											) }
-											shortcut={ shortcuts.insertBefore }
+											shortcut={ shortcuts.duplicate }
 										>
-											{ __( 'Add before' ) }
+											{ __( 'Duplicate' ) }
+										</MenuItem>
+									) }
+									{ canInsertBlock && ! isContentOnly && (
+										<>
+											<MenuItem
+												onClick={ pipe(
+													onClose,
+													onInsertBefore
+												) }
+												shortcut={
+													shortcuts.insertBefore
+												}
+											>
+												{ __( 'Add before' ) }
+											</MenuItem>
+											<MenuItem
+												onClick={ pipe(
+													onClose,
+													onInsertAfter
+												) }
+												shortcut={
+													shortcuts.insertAfter
+												}
+											>
+												{ __( 'Add after' ) }
+											</MenuItem>
+										</>
+									) }
+								</MenuGroup>
+								{ canCopyStyles && ! isContentOnly && (
+									<MenuGroup>
+										<CopyMenuItem
+											clientIds={ clientIds }
+											onCopy={ onCopy }
+											label={ __( 'Copy styles' ) }
+										/>
+										<MenuItem onClick={ onPasteStyles }>
+											{ __( 'Paste styles' ) }
 										</MenuItem>
+									</MenuGroup>
+								) }
+								<BlockSettingsMenuControls.Slot
+									fillProps={ {
+										onClose,
+										count,
+										firstBlockClientId,
+									} }
+									clientIds={ clientIds }
+								/>
```

Comment:
It looks like in some cases this slot will output the menu group, even when it's visually empty. To reproduce, if I use Edit mode and select a list item within TT4's blog home template, the settings menu will only include `Duplicate` as an option, but there'll be this extra empty area:

<img width="724" alt="image" src="https://github.com/user-attachments/assets/6a009448-6089-49ea-8fdd-a53b40694781">

<img width="604" alt="image" src="https://github.com/user-attachments/assets/66499dda-1f6a-4880-9b70-091aa995dc7b">


With a bit of commenting out, I think I've identified the three components that are rendering empty fills. They bail before outputting anything visually, however it results in `BlockSettingsMenuControlsSlot` thinking that [there are three fills here](https://github.com/WordPress/gutenberg/blob/cbcc28c5511dc87b81bca515b2e88fc1ec55e7e9/packages/block-editor/src/components/block-settings-menu-controls/index.js#L69-L75) and so the `MenuGroup` gets rendered which creates the extra space. Here's what I've commented out:

```diff
diff --git a/packages/editor/src/components/provider/index.js b/packages/editor/src/components/provider/index.js
index 11b1478d584..627e32e588f 100644
--- a/packages/editor/src/components/provider/index.js
+++ b/packages/editor/src/components/provider/index.js
@@ -327,17 +327,17 @@ export const ExperimentalEditorProvider = withRegistryProvider(
 							selection={ selection }
 							settings={ blockEditorSettings }
 							useSubRegistry={ false }
 						>
 							{ children }
 							{ ! settings.__unstableIsPreviewMode && (
 								<>
-									<PatternsMenuItems />
-									<TemplatePartMenuItems />
-									<ContentOnlySettingsMenu />
+									{/* <PatternsMenuItems /> */}
+									{/* <TemplatePartMenuItems /> */}
+									{/* <ContentOnlySettingsMenu /> */}
 									{ mode === 'template-locked' && (
 										<DisableNonPageContentBlocks />
 									) }
 									{ type === 'wp_navigation' && (
 										<NavigationBlockEditingMode />
 									) }
 									<EditorKeyboardShortcuts />
```

That then fixes up the spacing:

<img width="711" alt="image" src="https://github.com/user-attachments/assets/6492d98d-32d8-4896-bdc9-da408a2ee4d6">

<img width="644" alt="image" src="https://github.com/user-attachments/assets/2f30980e-d4e8-4b18-9d3a-3d5d97631c45">

So, in order to fix that spacing, do we need to conditionally render the above components, or update their internals so they don't output `<BlockSettingsMenuControls>` unexpectedly?

Initially I was wondering if we could fix it by using CSS with something like:

```css
.components-menu-group:has( > div:empty ) {
    display: none;
}
```

But unfortunately that still leaves a space due to the last menu group needing the negative margin in order for the spacing to look correct. And even with `display: none`, this empty group winds up getting that rule instead of the one that's visually there:

<img width="413" alt="image" src="https://github.com/user-attachments/assets/071c786e-c71c-4b7a-8216-9d8e2caa58ec">

So, unless there's another CSS-only way to handle it, we might need to lightly refactor some of those components that are outputting empty fills?

Reply:
Unfortunately, it's going to be very hard to solve these things by refactoring the fills. This is not the first item or the only place where we have this issue. I wonder if it's something fundamental to how slot/fills work.

If we can't find a CSS-only way for now, I think we should accept a small inconvenience for now.

Reply:
Ok I think I have a solution here but also, don't want to apply it in this PR though because it's potentially impactful. So I'm going to open a separate PR for that.

Reply:
This PR should fix it #65561